File: MyInventoryFoodCost.Core/Application/Services/MyInventoryFoodCost/uInventoryFoodCostBL.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;
using MyInventoryFoodCost.Core.Application.DTOs.Core;
using MyInventoryFoodCost.Core.Application.DTOs.DTOs;

namespace MyInventoryFoodCost.Core.Application.Services.MyInventoryFoodCost;

public class uInventoryFoodCostBL : IuInventoryFoodCostBL
{
    private readonly ILogger<uInventoryFoodCostBL> _logger;

    public uInventoryFoodCostBL(ILogger<uInventoryFoodCostBL> logger)
    {
        _logger = logger ?? throw new ArgumentNullException(nameof(logger));
    }

    /// <summary>
    /// Returns the full list of recipe costs with their ingredients.
    /// Legacy: GetRecipeCosts
    /// </summary>
    public List<TRecipeCost> GetRecipeCosts()
    {
        return BuildRecipeCosts();
    }

    /// <summary>
    /// Returns the ingredients for a specific recipe by searching through all recipe costs.
    /// Legacy: GetRecipeIngredients
    /// </summary>
    public List<TRecipeIngredient> GetRecipeIngredients(int recipeId)
    {
        var recipes = BuildRecipeCosts();

        for (int i = 0; i < recipes.Count; i++)
        {
            if (recipes[i].RecipeId == recipeId)
            {
                return recipes[i].Ingredients;
            }
        }

        // Recipe not found — return empty list (legacy returned nil)
        return new List<TRecipeIngredient>();
    }

    /// <summary>
    /// Returns cost trend data aggregated by period.
    /// Legacy: GetCostTrends
    /// </summary>
    public List<TCostTrend> GetCostTrends()
    {
        return BuildCostTrends();
    }

    /// <summary>
    /// Async wrapper for GetRecipeCosts.
    /// Legacy: GetRecipeCosts
    /// </summary>
    public async Task<List<TRecipeCost>> GetRecipeCostsAsync(CancellationToken ct = default)
    {
        await Task.CompletedTask;
        ct.ThrowIfCancellationRequested();

        return BuildRecipeCosts();
    }

    /// <summary>
    /// Async wrapper for GetRecipeIngredients.
    /// Legacy: GetRecipeIngredients
    /// </summary>
    public async Task<List<TRecipeIngredient>> GetRecipeIngredientsAsync(int recipeId, CancellationToken ct = default)
    {
        await Task.CompletedTask;
        ct.ThrowIfCancellationRequested();

        var recipes = BuildRecipeCosts();

        for (int i = 0; i < recipes.Count; i++)
        {
            if (recipes[i].RecipeId == recipeId)
            {
                return recipes[i].Ingredients;
            }
        }

        // Recipe not found — return empty list (legacy returned nil)
        return new List<TRecipeIngredient>();
    }

    /// <summary>
    /// Async wrapper for GetCostTrends.
    /// Legacy: GetCostTrends
    /// </summary>
    public async Task<List<TCostTrend>> GetCostTrendsAsync(CancellationToken ct = default)
    {
        await Task.CompletedTask;
        ct.ThrowIfCancellationRequested();

        return BuildCostTrends();
    }

    /// <summary>
    /// Builds the full list of recipe costs with their ingredients.
    /// Legacy: GetRecipeCosts
    /// </summary>
    private List<TRecipeCost> BuildRecipeCosts()
    {
        var result = new List<TRecipeCost>();

        // Recipe 1: Caesar Salad
        AddRecipe(result, 1, "Caesar Salad", "Appetizer", 3.75m, 12.50m,
            new List<TRecipeIngredient>
            {
                MakeIngredient("Romaine Lettuce", 0.25m, "kg", 3.00m),
                MakeIngredient("Parmesan", 0.05m, "kg", 28.00m),
                MakeIngredient("Croutons", 0.10m, "kg", 5.00m),
                MakeIngredient("Caesar Dressing", 0.05m, "liters", 12.00m)
            });

        // Recipe 2: Grilled Salmon
        AddRecipe(result, 2, "Grilled Salmon", "Main Course", 8.50m, 24.00m,
            new List<TRecipeIngredient>
            {
                MakeIngredient("Salmon Fillet", 0.25m, "kg", 22.00m),
                MakeIngredient("Olive Oil", 0.03m, "liters", 18.50m),
                MakeIngredient("Lemon", 0.05m, "kg", 4.00m),
                MakeIngredient("Asparagus", 0.15m, "kg", 8.00m),
                MakeIngredient("Butter", 0.03m, "kg", 7.50m)
            });

        // Recipe 3: Margherita Pizza
        AddRecipe(result, 3, "Margherita Pizza", "Main Course", 4.20m, 16.00m,
            new List<TRecipeIngredient>
            {
                MakeIngredient("Flour", 0.30m, "kg", 2.80m),
                MakeIngredient("Mozzarella", 0.15m, "kg", 12.00m),
                MakeIngredient("Tomatoes", 0.20m, "kg", 3.20m),
                MakeIngredient("Basil", 0.02m, "kg", 15.00m),
                MakeIngredient("Olive Oil", 0.02m, "liters", 18.50m)
            });

        // Recipe 4: Chicken Parmesan
        AddRecipe(result, 4, "Chicken Parmesan", "Main Course", 6.80m, 19.50m,
            new List<TRecipeIngredient>
            {
                MakeIngredient("Chicken Breast", 0.30m, "kg", 9.50m),
                MakeIngredient("Parmesan", 0.05m, "kg", 28.00m),
                MakeIngredient("Tomatoes", 0.15m, "kg", 3.20m),
                MakeIngredient("Mozzarella", 0.10m, "kg", 12.00m)
            });

        // Recipe 5: Risotto
        AddRecipe(result, 5, "Risotto", "Main Course", 5.10m, 18.00m,
            new List<TRecipeIngredient>
            {
                MakeIngredient("Rice", 0.20m, "kg", 4.50m),
                MakeIngredient("Parmesan", 0.06m, "kg", 28.00m),
                MakeIngredient("Butter", 0.05m, "kg", 7.50m),
                MakeIngredient("Chicken Breast", 0.15m, "kg", 9.50m),
                MakeIngredient("Heavy Cream", 0.08m, "liters", 5.50m)
            });

        // Recipe 6: Tiramisu
        AddRecipe(result, 6, "Tiramisu", "Dessert", 3.20m, 10.00m,
            new List<TRecipeIngredient>
            {
                MakeIngredient("Heavy Cream", 0.15m, "liters", 5.50m),
                MakeIngredient("Mascarpone", 0.10m, "kg", 14.00m),
                MakeIngredient("Espresso", 0.05m, "liters", 8.00m),
                MakeIngredient("Ladyfingers", 0.08m, "kg", 6.00m)
            });

        // Recipe 7: Bruschetta
        AddRecipe(result, 7, "Bruschetta", "Appetizer", 2.40m, 9.50m,
            new List<TRecipeIngredient>
            {
                MakeIngredient("Tomatoes", 0.15m, "kg", 3.20m),
                MakeIngredient("Basil", 0.02m, "kg", 15.00m),
                MakeIngredient("Garlic", 0.01m, "kg", 6.00m),
                MakeIngredient("Olive Oil", 0.03m, "liters", 18.50m)
            });

        // Recipe 8: Panna Cotta
        AddRecipe(result, 8, "Panna Cotta", "Dessert", 2.80m, 9.00m,
            new List<TRecipeIngredient>
            {
                MakeIngredient("Heavy Cream", 0.20m, "liters", 5.50m),
                MakeIngredient("Vanilla Extract", 0.01m, "liters", 45.00m),
                MakeIngredient("Sugar", 0.05m, "kg", 2.00m),
                MakeIngredient("Gelatin", 0.01m, "kg", 35.00m)
            });

        // Recipe 9: Seafood Pasta
        AddRecipe(result, 9, "Seafood Pasta", "Main Course", 9.20m, 22.00m,
            new List<TRecipeIngredient>
            {
                MakeIngredient("Pasta", 0.20m, "kg", 3.50m),
                MakeIngredient("Salmon Fillet", 0.15m, "kg", 22.00m),
                MakeIngredient("Shrimp", 0.10m, "kg", 28.00m),
                MakeIngredient("Garlic", 0.01m, "kg", 6.00m),
                MakeIngredient("Olive Oil", 0.03m, "liters", 18.50m)
            });

        // Recipe 10: Lemonade
        AddRecipe(result, 10, "Lemonade", "Beverage", 0.85m, 5.00m,
            new List<TRecipeIngredient>
            {
                MakeIngredient("Lemons", 0.10m, "kg", 4.00m),
                MakeIngredient("Sugar", 0.05m, "kg", 2.00m),
                MakeIngredient("Mint", 0.01m, "kg", 15.00m)
            });

        return result;
    }

    /// <summary>
    /// Builds the full list of cost trends.
    /// Legacy: GetCostTrends
    /// </summary>
    private List<TCostTrend> BuildCostTrends()
    {
        var trends = new List<TCostTrend>
        {
            new TCostTrend
            {
                Period = "2026-02",
                AvgCostPct = (double)(30.5m),
                TotalFoodCost = 12400.00m,
                TotalRevenue = 40655.74m
            },
            new TCostTrend
            {
                Period = "2026-03",
                AvgCostPct = (double)(31.2m),
                TotalFoodCost = 13100.00m,
                TotalRevenue = 41987.18m
            },
            new TCostTrend
            {
                Period = "2026-04",
                AvgCostPct = (double)(29.8m),
                TotalFoodCost = 12800.00m,
                TotalRevenue = 42953.02m
            },
            new TCostTrend
            {
                Period = "2026-05",
                AvgCostPct = (double)(32.1m),
                TotalFoodCost = 14200.00m,
                TotalRevenue = 44237.07m
            },
            new TCostTrend
            {
                Period = "2026-06",
                AvgCostPct = (double)(30.9m),
                TotalFoodCost = 13600.00m,
                TotalRevenue = 44012.94m
            },
            new TCostTrend
            {
                Period = "2026-07",
                AvgCostPct = (double)(31.5m),
                TotalFoodCost = 14500.00m,
                TotalRevenue = 46031.75m
            }
        };

        return trends;
    }

    /// <summary>
    /// Helper: creates a TRecipeIngredient with computed LineCost = Quantity * UnitCost.
    /// Legacy: MakeIngredient (nested function in GetRecipeCosts)
    /// </summary>
    private static TRecipeIngredient MakeIngredient(string name, decimal quantity, string unitMeasure, decimal unitCost)
    {
        // LineCost = Quantity * UnitCost, matching legacy: Result.LineCost := AQty * AUnitCost
        decimal lineCost = quantity * unitCost;

        return new TRecipeIngredient
        {
            IngredientName = name,
            Quantity = (double)(quantity),
            UnitMeasure = unitMeasure,
            UnitCost = unitCost,
            LineCost = lineCost
        };
    }

    /// <summary>
    /// Helper: adds a recipe to the list, computing CostPercentage and Profitable flag.
    /// Legacy: AddRecipe (nested procedure in GetRecipeCosts)
    /// </summary>
    private static void AddRecipe(
        List<TRecipeCost> recipes,
        int recipeId,
        string recipeName,
        string category,
        decimal totalCost,
        decimal sellingPrice,
        List<TRecipeIngredient> ingredients)
    {
        decimal costPercentage;
        if (sellingPrice > 0m)
        {
            // Legacy: (ACost / APrice) * 100
            costPercentage = Math.Round(totalCost / sellingPrice * 100m, 4, MidpointRounding.ToEven);
        }
        else
        {
            costPercentage = 0m;
        }

        // Legacy: Arr[Idx].Profitable := Arr[Idx].CostPercentage < 35
        bool profitable = costPercentage < 35m;

        recipes.Add(new TRecipeCost
        {
            RecipeId = recipeId,
            RecipeName = recipeName,
            Category = category,
            Ingredients = ingredients,
            TotalCost = totalCost,
            SellingPrice = sellingPrice,
            CostPercentage = (double)(costPercentage),
            Profitable = profitable
        });
    }
}

File: Delphi-source-code/MyInventory/MyInventoryFoodCost/src/uInventoryFoodCostBL.pas
unit uInventoryFoodCostBL;

interface

uses
  System.SysUtils;

type
  TRecipeIngredient = record
    IngredientName: string;
    Quantity: Double;
    UnitMeasure: string;
    UnitCost: Currency;
    LineCost: Currency;
  end;

  TRecipeCost = record
    RecipeId: Integer;
    RecipeName: string;
    Category: string;
    Ingredients: TArray<TRecipeIngredient>;
    TotalCost: Currency;
    SellingPrice: Currency;
    CostPercentage: Double;
    Profitable: Boolean;
  end;

  TCostTrend = record
    Period: string;
    AvgCostPct: Double;
    TotalFoodCost: Currency;
    TotalRevenue: Currency;
  end;

  TInventoryFoodCostBL = class
  public
    function GetRecipeCosts: TArray<TRecipeCost>;
    function GetRecipeIngredients(ARecipeId: Integer): TArray<TRecipeIngredient>;
    function GetCostTrends: TArray<TCostTrend>;
  end;

const
  SQL_GET_RECIPE_COSTS =
    'SELECT r.RecipeId, r.RecipeName, r.Category, ' +
    '       r.TotalCost, r.SellingPrice, ' +
    '       (r.TotalCost / r.SellingPrice * 100) AS CostPercentage ' +
    'FROM Recipes r ' +
    'ORDER BY r.Category, r.RecipeName';

  SQL_GET_RECIPE_INGREDIENTS =
    'SELECT ri.IngredientName, ri.Quantity, ri.UnitMeasure, ' +
    '       ri.UnitCost, (ri.Quantity * ri.UnitCost) AS LineCost ' +
    'FROM RecipeIngredients ri ' +
    'WHERE ri.RecipeId = :RecipeId ' +
    'ORDER BY ri.IngredientName';

  SQL_GET_COST_TRENDS =
    'SELECT FORMAT(ct.PeriodDate, ''yyyy-MM'') AS Period, ' +
    '       AVG(ct.CostPercentage) AS AvgCostPct, ' +
    '       SUM(ct.FoodCost) AS TotalFoodCost, ' +
    '       SUM(ct.Revenue) AS TotalRevenue ' +
    'FROM CostTrends ct ' +
    'GROUP BY FORMAT(ct.PeriodDate, ''yyyy-MM'') ' +
    'ORDER BY Period';

implementation

{ TInventoryFoodCostBL }

function TInventoryFoodCostBL.GetRecipeCosts: TArray<TRecipeCost>;

  function MakeIngredient(const AName: string; AQty: Double;
    const AUnit: string; AUnitCost: Currency): TRecipeIngredient;
  begin
    Result.IngredientName := AName;
    Result.Quantity := AQty;
    Result.UnitMeasure := AUnit;
    Result.UnitCost := AUnitCost;
    Result.LineCost := AQty * AUnitCost;
  end;

  procedure AddRecipe(var Arr: TArray<TRecipeCost>; AId: Integer;
    const AName, ACategory: string; ACost, APrice: Currency;
    const AIngredients: TArray<TRecipeIngredient>);
  var
    Idx: Integer;
  begin
    Idx := Length(Arr);
    SetLength(Arr, Idx + 1);
    Arr[Idx].RecipeId := AId;
    Arr[Idx].RecipeName := AName;
    Arr[Idx].Category := ACategory;
    Arr[Idx].Ingredients := AIngredients;
    Arr[Idx].TotalCost := ACost;
    Arr[Idx].SellingPrice := APrice;
    if APrice > 0 then
      Arr[Idx].CostPercentage := (ACost / APrice) * 100
    else
      Arr[Idx].CostPercentage := 0;
    Arr[Idx].Profitable := Arr[Idx].CostPercentage < 35;
  end;

begin
  Result := nil;

  AddRecipe(Result, 1, 'Caesar Salad', 'Appetizer', 3.75, 12.50,
    TArray<TRecipeIngredient>.Create(
      MakeIngredient('Romaine Lettuce', 0.25, 'kg', 3.00),
      MakeIngredient('Parmesan', 0.05, 'kg', 28.00),
      MakeIngredient('Croutons', 0.10, 'kg', 5.00),
      MakeIngredient('Caesar Dressing', 0.05, 'liters', 12.00)
    ));

  AddRecipe(Result, 2, 'Grilled Salmon', 'Main Course', 8.50, 24.00,
    TArray<TRecipeIngredient>.Create(
      MakeIngredient('Salmon Fillet', 0.25, 'kg', 22.00),
      MakeIngredient('Olive Oil', 0.03, 'liters', 18.50),
      MakeIngredient('Lemon', 0.05, 'kg', 4.00),
      MakeIngredient('Asparagus', 0.15, 'kg', 8.00),
      MakeIngredient('Butter', 0.03, 'kg', 7.50)
    ));

  AddRecipe(Result, 3, 'Margherita Pizza', 'Main Course', 4.20, 16.00,
    TArray<TRecipeIngredient>.Create(
      MakeIngredient('Flour', 0.30, 'kg', 2.80),
      MakeIngredient('Mozzarella', 0.15, 'kg', 12.00),
      MakeIngredient('Tomatoes', 0.20, 'kg', 3.20),
      MakeIngredient('Basil', 0.02, 'kg', 15.00),
      MakeIngredient('Olive Oil', 0.02, 'liters', 18.50)
    ));

  AddRecipe(Result, 4, 'Chicken Parmesan', 'Main Course', 6.80, 19.50,
    TArray<TRecipeIngredient>.Create(
      MakeIngredient('Chicken Breast', 0.30, 'kg', 9.50),
      MakeIngredient('Parmesan', 0.05, 'kg', 28.00),
      MakeIngredient('Tomatoes', 0.15, 'kg', 3.20),
      MakeIngredient('Mozzarella', 0.10, 'kg', 12.00)
    ));

  AddRecipe(Result, 5, 'Risotto', 'Main Course', 5.10, 18.00,
    TArray<TRecipeIngredient>.Create(
      MakeIngredient('Rice', 0.20, 'kg', 4.50),
      MakeIngredient('Parmesan', 0.06, 'kg', 28.00),
      MakeIngredient('Butter', 0.05, 'kg', 7.50),
      MakeIngredient('Chicken Breast', 0.15, 'kg', 9.50),
      MakeIngredient('Heavy Cream', 0.08, 'liters', 5.50)
    ));

  AddRecipe(Result, 6, 'Tiramisu', 'Dessert', 3.20, 10.00,
    TArray<TRecipeIngredient>.Create(
      MakeIngredient('Heavy Cream', 0.15, 'liters', 5.50),
      MakeIngredient('Mascarpone', 0.10, 'kg', 14.00),
      MakeIngredient('Espresso', 0.05, 'liters', 8.00),
      MakeIngredient('Ladyfingers', 0.08, 'kg', 6.00)
    ));

  AddRecipe(Result, 7, 'Bruschetta', 'Appetizer', 2.40, 9.50,
    TArray<TRecipeIngredient>.Create(
      MakeIngredient('Tomatoes', 0.15, 'kg', 3.20),
      MakeIngredient('Basil', 0.02, 'kg', 15.00),
      MakeIngredient('Garlic', 0.01, 'kg', 6.00),
      MakeIngredient('Olive Oil', 0.03, 'liters', 18.50)
    ));

  AddRecipe(Result, 8, 'Panna Cotta', 'Dessert', 2.80, 9.00,
    TArray<TRecipeIngredient>.Create(
      MakeIngredient('Heavy Cream', 0.20, 'liters', 5.50),
      MakeIngredient('Vanilla Extract', 0.01, 'liters', 45.00),
      MakeIngredient('Sugar', 0.05, 'kg', 2.00),
      MakeIngredient('Gelatin', 0.01, 'kg', 35.00)
    ));

  AddRecipe(Result, 9, 'Seafood Pasta', 'Main Course', 9.20, 22.00,
    TArray<TRecipeIngredient>.Create(
      MakeIngredient('Pasta', 0.20, 'kg', 3.50),
      MakeIngredient('Salmon Fillet', 0.15, 'kg', 22.00),
      MakeIngredient('Shrimp', 0.10, 'kg', 28.00),
      MakeIngredient('Garlic', 0.01, 'kg', 6.00),
      MakeIngredient('Olive Oil', 0.03, 'liters', 18.50)
    ));

  AddRecipe(Result, 10, 'Lemonade', 'Beverage', 0.85, 5.00,
    TArray<TRecipeIngredient>.Create(
      MakeIngredient('Lemons', 0.10, 'kg', 4.00),
      MakeIngredient('Sugar', 0.05, 'kg', 2.00),
      MakeIngredient('Mint', 0.01, 'kg', 15.00)
    ));
end;

function TInventoryFoodCostBL.GetRecipeIngredients(ARecipeId: Integer): TArray<TRecipeIngredient>;
var
  Recipes: TArray<TRecipeCost>;
  I: Integer;
begin
  Result := nil;
  Recipes := GetRecipeCosts;
  for I := 0 to High(Recipes) do
  begin
    if Recipes[I].RecipeId = ARecipeId then
    begin
      Result := Recipes[I].Ingredients;
      Exit;
    end;
  end;
end;

function TInventoryFoodCostBL.GetCostTrends: TArray<TCostTrend>;

  procedure AddTrend(var Arr: TArray<TCostTrend>; const APeriod: string;
    AAvgPct: Double; AFoodCost, ARevenue: Currency);
  var
    Idx: Integer;
  begin
    Idx := Length(Arr);
    SetLength(Arr, Idx + 1);
    Arr[Idx].Period := APeriod;
    Arr[Idx].AvgCostPct := AAvgPct;
    Arr[Idx].TotalFoodCost := AFoodCost;
    Arr[Idx].TotalRevenue := ARevenue;
  end;

begin
  Result := nil;
  AddTrend(Result, '2026-02', 30.5, 12400.00, 40655.74);
  AddTrend(Result, '2026-03', 31.2, 13100.00, 41987.18);
  AddTrend(Result, '2026-04', 29.8, 12800.00, 42953.02);
  AddTrend(Result, '2026-05', 32.1, 14200.00, 44237.07);
  AddTrend(Result, '2026-06', 30.9, 13600.00, 44012.94);
  AddTrend(Result, '2026-07', 31.5, 14500.00, 46031.75);
end;

end.
